File: leetcode/js/count-number-of-maximum-bitwise-or-subsets.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var countMaxOrSubsets = function (nums) {
  let subsets = [[]];

  for (const num of nums) {
    const newSubsets = subsets.map((subset) => [...subset, num]);
    subsets = [...subsets, ...newSubsets];
  }

  const max = nums.reduce((acc, cur) => acc | cur, 0);

  return subsets.filter(
    (subset) => subset.reduce((acc, cur) => acc | cur, 0) === max
  ).length;
};

var countMaxOrSubsets = function (nums) {
  let result = 0;
  const maxOr = nums.reduce((acc, cur) => acc | cur, 0);
  const dfs = (idx, sum) => {
    if (sum === maxOr) {
      result++;
    }

    if (idx === nums.length) {
      return;
    }

    for (let i = idx; i < nums.length; i++) {
      dfs(i + 1, sum | nums[i]);
    }
  };

  dfs(0, 0);

  return result;
};

File: leetcode/rust/count-number-of-maximum-bitwise-or-subsets.rs
impl Solution1 {
    pub fn count_max_or_subsets(nums: Vec<i32>) -> i32 {
        let mut result = 0;
        let max_or = nums.iter().fold(0, |acc, cur| acc | cur);

        Self::dfs(0, 0, &nums, max_or, &mut result);

        result
    }

    fn dfs(idx: usize, sum: i32, nums: &Vec<i32>, max_or: i32, result: &mut i32) {
        if sum == max_or {
            *result += 1;
        }

        for i in idx..nums.len() {
            Self::dfs(i + 1, sum | nums[i], nums, max_or, result);
        }
    }
}

impl Solution2 {
    pub fn count_max_or_subsets(nums: Vec<i32>) -> i32 {
        let max = nums.iter().fold(0, |acc, cur| acc | cur);

        Self::solve(&nums, 0, max)
    }

    fn solve(nums: &[i32], or: i32, max: i32) -> i32 {
        match nums {
            [n, rest @ ..] if or | n == or => Self::solve(rest, or, max) * 2,
            [n, rest @ ..] => Self::solve(rest, or, max) + Self::solve(rest, or | n, max),
            [] => {
                if or == max {
                    1
                } else {
                    0
                }
            }
        }
    }
}
